### PR TITLE
[multistage] Push Down Project for Aggregation With No Grouping Sets

### PR DESCRIPTION
**Pull request by ankitsultana**
Issue has details regarding the problem: https://github.com/apache/pinot/issues/10129

Open to feedback or any alternate approaches.

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotRuleUtils.java [hunk 1]
```diff
@@ -23,22 +23,28 @@
...
   private static final RelBuilder.Config DEFAULT_CONFIG =
       RelBuilder.Config.DEFAULT.withAggregateUnique(true).withPushJoinCondition(true);
 
-  public static final RelBuilderFactory PINOT_REL_FACTORY =
+  static final RelBuilderFactory PINOT_REL_FACTORY =
       RelBuilder.proto(Contexts.of(RelFactories.DEFAULT_STRUCT, DEFAULT_CONFIG));
 
+  // A AggregateExtractProjectRule configured to run when a LogicalAggregate sits atop a LogicalJoin.
+  static final AggregateExtractProjectRule PINOT_AGGREGATE_EXTRACT_PROJECT_RULE =
+      new AggregateExtractProjectRule(LogicalAggregate.class, LogicalJoin.class, PinotRuleUtils.PINOT_REL_FACTORY);
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotRuleUtils.java:41]:
```suggestion
  static final AggregateExtractProjectRule PINOT_AGGREGATE_EXTRACT_PROJECT_RULE =
      new AggregateExtractProjectRule(ImmutableAggregateExtractProjectRule.Config.of()
          .withOperandSupplier(b0 -> b0.operand(Aggregate.class).oneInput(
              b1 -> b1.operand(RelNode.class).predicate(r -> !(r instanceof Project)).anyInputs()))
          .withRelBuilderFactory(PINOT_REL_FACTORY));
```

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotQueryRuleSets.java [hunk 2]
```diff
@@ -43,6 +43,8 @@ private PinotQueryRuleSets() {
           CoreRules.FILTER_AGGREGATE_TRANSPOSE,
           // push filter through set operation
           CoreRules.FILTER_SET_OP_TRANSPOSE,
+          // create a project on top of an aggregate when the aggregate is on top of a join
+          PinotRuleUtils.PINOT_AGGREGATE_EXTRACT_PROJECT_RULE,
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotQueryRuleSets.java:47]: this approach doesn't work on count(*) or count(1) as 
1. pinot datatypes are non-nullable for now (which I think we should fix)
2. COUNT(*) and COUNT(1) doesn't have operand so it cannot be extracted into a project, we might need a special rule unfortunately.

**Comment by ankitsultana** [comment]
Superseded by https://github.com/apache/pinot/pull/10156

**Comment by walterddr** [comment]
:+1:

**Pull request closed without merging**